### PR TITLE
Allow the Input/EditableText widget to scroll horizontally

### PR DESCRIPTION
**Pull request by jason-simmons**
EditableText is now rendered using a custom RenderObject
(RenderEditableParagraph).  RenderEditableParagraph draws the cursor,
handles scroll offsets, and provides feedback about the size of the text for
use by the scroll behavior.

### CODE REVIEW

**Comment by jason-simmons** [comment]
Some questions:

* Should RenderEditableParagraph be a subclass of RenderParagraph, or would it be cleaner to fork RenderParagraph?

* Does the adjustment for the scroll offset in RenderEditableParagraph.paint look correct?
I saw some subtleties around this in RenderViewport and wasn't sure if RenderEditableParagraph needed to do something similar.

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+  Color _cursorColor;
+  bool _showCursor;
+  SizeChangedCallback onContentSizeChanged;
+  Offset _scrollOffset;
+
+  Size _contentSize;
+
+  Color get cursorColor => _cursorColor;
+  void set cursorColor(Color cursorColor) {
+    _cursorColor = cursorColor;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: In these setters, we usually check whether the old value is equal to the new value and skip the markNeedsPaint call in that case.  We do that because the widget layer generates unconditional writes into the render tree and we don't want to mark ourselves as needing paint when its not actually needed.

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    markNeedsPaint();
+  }
+
+  bool get showCursor => _showCursor;
+  void set showCursor(bool showCursor) {
+    _showCursor = showCursor;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: ditto

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    markNeedsPaint();
+  }
+
+  Offset get scrollOffset => _scrollOffset;
+  void set scrollOffset(Offset scrollOffset) {
+    _scrollOffset = scrollOffset;
+    markNeedsPaint();
+  }
+
+  bool allowLineWrap() => false;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: `bool get allowLineWrap => false;`

That way it's a property (more idomatic)

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        onContentSizeChanged(newContentSize);
+      }
+    }
+  }
+
+  void paint(PaintingContext context, Offset offset) {
+    layoutText(constraints);
+
+    context.canvas.save();
+    context.canvas.clipRect(offset & size);
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: We should skip the clip rect if the _contentSize.width <= size.width.  The clip generates costs layer in the pipeline.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 5]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
 // differently. To work around this problem for now, we round fractional pixel
 // values up to the nearest whole pixel value. The right long-term fix is to do
 // layout using fixed precision arithmetic.
-double _applyFloatingPointHack(double layoutValue) {
+double applyFloatingPointHack(double layoutValue) {
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Oh, it's sad to make this public.  Maybe we should push this logic down into TextPainter.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 6]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
...
 class RenderParagraph extends RenderBox {
 
-  RenderParagraph(TextSpan text) : _textPainter = new TextPainter(text) {
+  RenderParagraph(TextSpan text) : textPainter = new TextPainter(text) {
     assert(text != null);
   }
 
-  TextPainter _textPainter;
+  // Exposed for use by subclasses.
+  TextPainter textPainter;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: final ?

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 7]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
...
 /// A render object that displays a paragraph of text
 class RenderParagraph extends RenderBox {
 
-  RenderParagraph(TextSpan text) : _textPainter = new TextPainter(text) {
+  RenderParagraph(TextSpan text) : textPainter = new TextPainter(text) {
     assert(text != null);
   }
 
-  TextPainter _textPainter;
+  // Exposed for use by subclasses.
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: I'd skip comments like these given that there's no way to enforce them.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 8]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
...
+    if (textPainter.text == value)
       return;
-    _textPainter.text = value;
+    textPainter.text = value;
     _constraintsForCurrentLayout = null;
     markNeedsLayout();
   }
 
-  void _layout(BoxConstraints constraints) {
+  // Can be overridden by subclasses.
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: I'd skip this comment too.  Everything can be overridden by subclasses.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 9]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
...
+  bool allowLineWrap() => true;
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Why not take this value as a constructor parameter and provide a getter and a setter?  As written, the value is neither final nor is there a notification pathway for when it changes.  We should either make it final (e.g., by taking the value as a constructor parameter and not providing a setter) or have a notification pathway for it changing (e.g., by having a setting that calls markNeedsLayout as appropriate).

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 10]
```diff
@@ -15,59 +15,64 @@ export 'package:sky/src/painting/text_painter.dart';
...
+
+  // Exposed for use by subclasses.
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: I'd skip these comments.

Rather than exposing all these implementation details for `RenderParagraph`, maybe we should push this logic down into TextPainter.  It would be sensible for TextPainter's layout function to work based on a BoxConstraints rather than having separate values for each of the fields on BoxConstraints.  Depending on where BoxConstraints is implemented, we might need to push that whole object down into the painting layer, but that's fine.

File: sky/packages/sky/lib/src/fn3/editable_text.dart [hunk 11]
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
...
 
-    Widget text;
+  final EditableString value;
+  final TextStyle style;
+  final Color cursorColor;
+  final bool showCursor;
+  final SizeChangedCallback onContentSizeChanged;
+  final Offset scrollOffset;
+
+  RenderEditableParagraph createRenderObject() => new RenderEditableParagraph(
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/editable_text.dart]: style nits:

- only use => if the expression fits on the same line. Otherwise, use a block with return.
- If you have a newline after a (, {, or [, you should have a newline before the corresponding ), }, or ]. (In fact in general, whitespace around open/close punctuation should be symmetric.)
- Indent two spaces from the previous indent.

File: sky/packages/sky/lib/src/fn3/editable_text.dart [hunk 12]
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
...
+  TextSpan _buildTextSpan() {
     if (value.composing.isValid) {
-      TextStyle composingStyle = style.merge(const TextStyle(decoration: underline));
-      text = new StyledText(elements: [
-        style,
-        value.textBefore(value.composing),
-        [composingStyle, value.textInside(value.composing)],
-        value.textAfter(value.composing)
+      TextStyle composingStyle = style.merge(
+          const TextStyle(decoration: underline));
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/editable_text.dart]: here's another example where you'd want to put the ) on a new line

File: sky/packages/sky/lib/src/fn3/editable_text.dart [hunk 13]
```diff
@@ -207,29 +195,67 @@ class EditableTextState extends State<EditableText> {
...
+
+      return new StyledTextSpan(style, [
+          new PlainTextSpan(value.textBefore(value.composing)),
+          new StyledTextSpan(composingStyle, [
+              new PlainTextSpan(value.textInside(value.composing))
+          ]),
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/editable_text.dart]: this is good style

File: sky/packages/sky/lib/src/fn3/input.dart [hunk 14]
```diff
@@ -18,14 +21,16 @@ typedef void StringValueChanged(String value);
...
-class Input extends StatefulComponent {
+class Input extends Scrollable {
   Input({
     GlobalKey key,
     this.initialValue: '',
     this.placeholder,
     this.onChanged,
     this.keyboardType: KeyboardType.TEXT
-  }): super(key: key);
+  }): super(key: key,
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/input.dart]: put a newline after the (, then indent two spaces for the arguments, then ) on a line aligned with the }.

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart (hunk 1 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: Use "value" or "newCursorColor" to avoid shadowing "cursorColor" here. Generally I prefer "value" because it makes copy/pasting setters easier later.

File: sky/packages/sky/lib/src/rendering/paragraph.dart (hunk 10 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: BoxConstraints extends Constraints which is defined in object.dart, but it doesn't currently reference RenderObject in any way, so I guess it could be pushed to painting/ safely. If you do this, put it in a new file, add it to painting,dart, and check the examples and tests to make sure anyone who uses it now imports painting.dart. (Do this in a separate PR if you do it.)

**Comment by jason-simmons** [comment]
Addressed code review feedback - PTAL.

I made RenderParagraph._allowLineWrap final.  allowLineWrap was added so that RenderEditableParagraph could force text to render on a single line and then apply clipping.  I didn't envision it as a mutable property with a public API.

File: sky/packages/sky/lib/src/fn3/editable_text.dart [hunk 15]
```diff
@@ -207,29 +196,72 @@ class EditableTextState extends State<EditableText> {
...
+      );
+
+      return new StyledTextSpan(style, [
+        new PlainTextSpan(value.textBefore(value.composing)),
+        new StyledTextSpan(composingStyle, [
+          new PlainTextSpan(value.textInside(value.composing))
+        ]),
+        new PlainTextSpan(value.textAfter(value.composing))
       ]);
     } else {
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/editable_text.dart]: No need for the else here, since the previous block returned.

File: sky/packages/sky/lib/src/fn3/input.dart [hunk 16]
```diff
@@ -100,12 +112,17 @@ class InputState extends State<Input> {
...
       decoration: new BoxDecoration(border: focusHighlight)
     );
 
+    SizeObserver input = new SizeObserver(
+      child: container,
+      callback: _handleContainerSizeChanged
+    );
+
     return new Listener(
       child: input,
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/input.dart]: I'd just inline container and input into the return statement here. Not sure why it was done separately before. It's easier to visualise the tree if it's all inline.

File: sky/packages/sky/lib/src/rendering/paragraph.dart (hunk 5 above)
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Either push the logic that uses applyFloatingPointHack into TextPainter as per adam's comment, or file a bug and mention it in the checkin comment.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 17]
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
...
+double applyFloatingPointHack(double layoutValue) {
   return layoutValue.ceilToDouble();
 }
 
 /// A render object that displays a paragraph of text
 class RenderParagraph extends RenderBox {
 
-  RenderParagraph(TextSpan text) : _textPainter = new TextPainter(text) {
+  RenderParagraph(
+      TextSpan text,
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Indent two characters, not four. Put spaces around the named arguments.

(See the style guide for details.)

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 18]
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
...
+  ) : textPainter = new TextPainter(text),
+      _allowLineWrap = allowLineWrap {
     assert(text != null);
   }
 
-  TextPainter _textPainter;
+  final TextPainter textPainter;
+
+  // Whether the text should be allowed to wrap to multiple lines.
+  final bool _allowLineWrap;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Why is this an immutable field? I think it's more idiomatic to use a getter.
That is, have this here:

```dart
   bool get allowLineWrap => true;
```

...and then override it in the subclass to return false.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 19]
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
...
+    textPainter.layout();
     _constraintsForCurrentLayout = constraints;
   }
 
   double getMinIntrinsicWidth(BoxConstraints constraints) {
-    _layout(constraints);
+    layoutText(constraints);
     return constraints.constrainWidth(
-        _applyFloatingPointHack(_textPainter.minContentWidth));
+        applyFloatingPointHack(textPainter.minContentWidth));
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: Put this all on one line or put the closing ) on its own line.

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 20]
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
...
   }
 
   double getMaxIntrinsicWidth(BoxConstraints constraints) {
-    _layout(constraints);
+    layoutText(constraints);
     return constraints.constrainWidth(
-        _applyFloatingPointHack(_textPainter.maxContentWidth));
+        applyFloatingPointHack(textPainter.maxContentWidth));
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: ditto

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 21]
```diff
@@ -15,59 +15,66 @@ export 'package:sky/src/painting/text_painter.dart';
...
   }
 
   double _getIntrinsicHeight(BoxConstraints constraints) {
-    _layout(constraints);
+    layoutText(constraints);
     return constraints.constrainHeight(
-        _applyFloatingPointHack(_textPainter.height));
+        applyFloatingPointHack(textPainter.height));
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: ditto

File: sky/packages/sky/lib/src/rendering/paragraph.dart [hunk 22]
```diff
@@ -80,15 +87,15 @@ class RenderParagraph extends RenderBox {
...
-    _layout(constraints);
-    return _textPainter.computeDistanceToActualBaseline(baseline);
+    layoutText(constraints);
+    return textPainter.computeDistanceToActualBaseline(baseline);
   }
 
   void performLayout() {
-    _layout(constraints);
+    layoutText(constraints);
     // _paragraphPainter.width always expands to fill, use maxContentWidth instead.
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/paragraph.dart]: replace this comment with "We use textPainter.maxContentWidth here, rather that textPainter.width, because the latter is the width that it used to wrap the text, whereas the former is the actual width of the text."

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 23]
```diff
@@ -0,0 +1,115 @@
...
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:sky/painting.dart';
+import 'package:sky/src/rendering/box.dart';
+import 'package:sky/src/rendering/object.dart';
+import 'package:sky/src/rendering/paragraph.dart';
+import 'package:sky/src/rendering/proxy_box.dart' show SizeChangedCallback;
+
+const _kCursorGap = 1.0;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: put comments after these constants documenting the units, e.g.:

`const _kCursorGap = 1.0; // pixels`

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 24]
```diff
@@ -0,0 +1,115 @@
...
+const _kCursorHeightOffset = 2.0;
+const _kCursorWidth = 1.0;
+
+/// A render object used by EditableText widgets.  This is similar to
+/// RenderParagraph but also renders a cursor and provides support for
+/// scrolling.
+class RenderEditableParagraph extends RenderParagraph {
+
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: If you have two newlines after the {, put two newlines before the } at the end of the class.

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 25]
```diff
@@ -0,0 +1,115 @@
...
+  void set showCursor(bool value) {
+    if (_showCursor == value)
+      return;
+    _showCursor = showCursor;
+    markNeedsPaint();
+  }
+
+  Offset get scrollOffset => _scrollOffset;
+  void set scrollOffset(Offset scrollOffset) {
+    _scrollOffset = scrollOffset;
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: check that it hasn't changed

File: sky/packages/sky/lib/src/fn3/input.dart [hunk 26]
```diff
@@ -125,4 +142,26 @@ class InputState extends State<Input> {
...
+    _updateScrollBehavior();
+  }
+
+  void _updateScrollBehavior() {
+    // Set the scroll offset to match the content width so that the cursor
+    // (which is always at the end of the text) will be visible.
+    scrollTo(scrollBehavior.updateExtents(
+      contentExtent: _contentWidth,
+      containerExtent: _containerWidth,
+      scrollOffset: _contentWidth));
```
Comment (Hixie) [inline sky/packages/sky/lib/src/fn3/input.dart]: newline before last ), per style guide

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+    markNeedsPaint();
+  }
+
+  double getMinIntrinsicWidth(BoxConstraints constraints) {
+    return constraints.constrainWidth(
+      super.getMinIntrinsicWidth(constraints) + _kCursorGap + _kCursorWidth
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: This isn't accurate. For example, consider the case where the text looks like this:

```
Hello Nice Big
World|
```

...where the cursor is the vertical bar. The minimum intrinsic width in this case is unaffected by the cursor.

If you only support single-line text, such that there's only one intrinsic width, then just have a single function that you call from both getMinIntrinsicWidth and getMaxIntrinsicWidth (and assert that textPainter.minContentWidth and textPainter.maxContentWidth are the same, and document that this is only valid because we don't wrap).

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 28]
```diff
@@ -0,0 +1,115 @@
...
+
+    Size newContentSize = new Size(
+      applyFloatingPointHack(textPainter.maxContentWidth) + _kCursorGap + _kCursorWidth,
+      applyFloatingPointHack(textPainter.height)
+    );
+    size = constraints.constrain(newContentSize);
+
+    if (_contentSize == null || _contentSize != newContentSize) {
+      _contentSize = newContentSize;
+      if (onContentSizeChanged != null) {
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: remove the { } around the block if it's a one-line block.

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 29]
```diff
@@ -0,0 +1,115 @@
...
+      }
+    }
+  }
+
+  void paint(PaintingContext context, Offset offset) {
+    layoutText(constraints);
+
+    context.canvas.save();
+
+    if (_contentSize.width > size.width) {
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: remove braces around one line blocks

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+      context.canvas.clipRect(offset & size);
+    }
+
+    textPainter.paint(context.canvas, offset - _scrollOffset);
+
+    if (_showCursor) {
+      Rect cursorRect =  new Rect.fromLTWH(
+        textPainter.maxContentWidth + _kCursorGap,
```
Comment (Hixie) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: shouldn't you be applying the floating point hack here?

**Comment by Hixie** [comment]
Mostly just style nits.

**Comment by jason-simmons** [comment]
Updated the CL

* Moved applyFloatingPointHack rounding to the (min|max)ContentWidth and height properties in TextPainter
* Skipped the (minContentWidth == maxContentWidth) check in EditableParagraph.getIntrinsicWidth.  minContentWidth reflects the width if the text was wrapped at every point even if the TextPainter's maxWidth is set to infinity

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart [hunk 31]
```diff
@@ -0,0 +1,124 @@
...
+      _contentSize = newContentSize;
+      if (onContentSizeChanged != null)
+        onContentSizeChanged(newContentSize);
+    }
+  }
+
+  void paint(PaintingContext context, Offset offset) {
+    layoutText(constraints);
+
+    context.canvas.save();
```
Comment (abarth) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: We only need to call save/restore when we call clipRect.  We should move those calls inside the conditional too.

**Comment by abarth** [comment]
LGTM

File: sky/packages/sky/lib/src/rendering/editable_paragraph.dart (hunk 31 above)
Comment (jason-simmons) [inline sky/packages/sky/lib/src/rendering/editable_paragraph.dart]: Done

**Comment by Hixie** [comment]
LGTM
